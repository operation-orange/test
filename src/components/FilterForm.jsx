import React from 'react';
import PropTypes from 'prop-types';
import connectState from './FilterForm.state';

const changeHandler = (action, field) => e => action({ [field]: e.currentTarget.value });

export const FilterForm = ({
  name, rating, facility, updateFilterFormAction,
}) => (
  <div className="filter-form">
    <input
      type="text"
      className="filter-form__name-field"
      onChange={changeHandler(updateFilterFormAction, 'name')}
      value={name}
    />
    <input
      type="text"
      className="filter-form__rating-field"
      onChange={changeHandler(updateFilterFormAction, 'rating')}
      value={rating}
    />
    <input
      type="text"
      className="filter-form__facility-field"
      onChange={changeHandler(updateFilterFormAction, 'facility')}
      value={facility}
    />
    <button type="button">Filter</button>
  </div>
);

FilterForm.propTypes = {
  name: PropTypes.string.isRequired,
  rating: PropTypes.string.isRequired,
  facility: PropTypes.string.isRequired,
  updateFilterFormAction: PropTypes.func.isRequired,
};

export default connectState(FilterForm);
