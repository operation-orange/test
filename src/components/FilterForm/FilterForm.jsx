import React from 'react';
import PropTypes from 'prop-types';
import connectState from './FilterForm.state';

const textChangeHandler = (action, field) => e => action({ [field]: e.target.value });
const facilitiesChangeHandler = (action, selectedFacilities) => (e) => {
  if (e.target.checked) {
    action({ facilities: [...selectedFacilities, e.target.value] });
  } else {
    action({ facilities: selectedFacilities.filter(facility => facility !== e.target.value) });
  }
};

export const FilterForm = ({
  name, rating, selectedFacilities, updateFilterFormAction, filterHotelsAction, availableFacilities,
}) => (
  <div className="filter-form">
    Hotel Name:
    {' '}
    <input
      type="text"
      className="filter-form__name-field"
      onChange={textChangeHandler(updateFilterFormAction, 'name')}
      value={name}
    />
    {' '}
    Rating:
    {' '}
    <input
      type="text"
      className="filter-form__rating-field"
      onChange={textChangeHandler(updateFilterFormAction, 'rating')}
      value={rating}
    />
    {' '}
    Facility:
    {' '}
    {availableFacilities.map((facility, i) => (
      <label key={facility} htmlFor={`facility${i}`}>
        {facility}
        <input
          id={`facility${i}`}
          type="checkbox"
          value={facility}
          onChange={facilitiesChangeHandler(updateFilterFormAction, selectedFacilities)}
          checked={selectedFacilities.includes(facility)}
        />
      </label>
    ))}
    <button
      type="button"
      onClick={filterHotelsAction}
    >
      Filter
    </button>
  </div>
);

FilterForm.propTypes = {
  name: PropTypes.string.isRequired,
  rating: PropTypes.string.isRequired,
  selectedFacilities: PropTypes.arrayOf(PropTypes.string).isRequired,
  updateFilterFormAction: PropTypes.func.isRequired,
  filterHotelsAction: PropTypes.func.isRequired,
  availableFacilities: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default connectState(FilterForm);
