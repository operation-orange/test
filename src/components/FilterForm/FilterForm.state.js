import { connect } from 'react-redux';

import { updateFilterForm, filterHotels } from '../../state/hotels/actions';

export const mapStateToProps = ({ hotels: { filterForm, availableFacilities } }) => ({
  name: filterForm.name,
  rating: filterForm.rating,
  selectedFacilities: filterForm.facilities,
  availableFacilities,
});

export const mapDispatchToProps = {
  updateFilterFormAction: updateFilterForm,
  filterHotelsAction: filterHotels,
};

const connectState = Component => connect(
  mapStateToProps,
  mapDispatchToProps,
)(Component);

export default connectState;
