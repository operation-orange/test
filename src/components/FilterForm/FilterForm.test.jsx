import { shallow } from 'enzyme';
import React from 'react';
import { FilterForm } from './FilterForm';

test('renders default state passing form data', () => {
  const wrapper = shallow(
    <FilterForm
      name="the name"
      rating="the rating"
      facility="the facility"
      selectedFacilities={[]}
      availableFacilities={[]}
      updateFilterFormAction={() => {}}
      filterHotelsAction={() => {}}
    />,
  );
  expect(wrapper).toMatchSnapshot();
});

test.each(['name', 'rating'])('onChange for the %s field triggers updateFilterFormAction', (field) => {
  const mockUpdateFilterFormAction = jest.fn();
  const wrapper = shallow(
    <FilterForm
      name="the name"
      rating="the rating"
      facility="the facility"
      selectedFacilities={[]}
      availableFacilities={[]}
      updateFilterFormAction={mockUpdateFilterFormAction}
      filterHotelsAction={() => {}}
    />,
  );

  wrapper.find(`.filter-form__${field}-field`).simulate('change', {
    target: {
      value: 'test value',
    },
  });

  expect(mockUpdateFilterFormAction).toHaveBeenCalledWith({
    [field]: 'test value',
  });
});

// TODO: facility filter test
