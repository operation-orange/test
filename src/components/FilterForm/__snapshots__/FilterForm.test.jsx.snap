// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default state passing form data 1`] = `
<div
  className="filter-form"
>
  Hotel Name:
   
  <input
    className="filter-form__name-field"
    onChange={[Function]}
    type="text"
    value="the name"
  />
   
  Rating:
   
  <input
    className="filter-form__rating-field"
    onChange={[Function]}
    type="text"
    value="the rating"
  />
   
  Facility:
   
  <button
    onClick={[Function]}
    type="button"
  >
    Filter
  </button>
</div>
`;
